### PR TITLE
fix: solve multiple rerun problem

### LINKED ISSUE #149: bug: git checkout error
Поправил генерацию, но столкнулся с этой проблемой:

Image

Надеюсь у нас так и задумано на данный момент времени

### LINKED ISSUE #165: bug: cannot upload PDF report
## Repository

https://github.com/LISA-ITMO/SchoolOCR 

## Logs

```
[23:18:43] INFO     [23:18:43] - Config successfully updated and loaded                                                                                                  run.py:246
           INFO     [23:18:43] - Successfully fetched metadata for repository: https://github.com/LISA-ITMO/SchoolOCR                                               metadata.py:115
[23:18:44] INFO     [23:18:44] - Repository LISA-ITMO/SchoolOCR is already starred.                                                                              github_agent.py:94
           INFO     [23:18:44] - Fork created successfully: https://github.com/osa-bot/SchoolOCR                                                                 github_agent.py:69
           INFO     [23:18:44] - Cloning the main branch from https://github.com/LISA-ITMO/SchoolOCR into directory /home/ilya/OSA/SchoolOCR...                 github_agent.py:134
[23:24:57] INFO     [23:24:57] - Cloning completed                                                                                                              github_agent.py:141
           INFO     [23:24:57] - Creating and switching to branch osa_tool...                                                                                   github_agent.py:162
           INFO     [23:24:57] - Switched to branch osa_tool.                                                                                                   github_agent.py:164
           INFO     [23:24:57] - Starting analysis for repository LISA-ITMO/SchoolOCR                                                                           report_maker.py:334
[23:25:22] INFO     [23:25:22] - <Response [200]>                                                                                                                     models.py:193
           INFO     [23:25:22] - PDF report successfully created in /home/ilya/OSA/SchoolOCR/SchoolOCR_report.pdf                                               report_maker.py:354
           INFO     [23:25:22] - Creating and switching to branch osa_tool_attachments...                                                                       github_agent.py:162
           INFO     [23:25:22] - Switched to branch osa_tool_attachments.                                                                                       github_agent.py:164
           INFO     [23:25:22] - Committing changes...                                                                                                          github_agent.py:183
           ERROR    [23:25:22] - Error: Cmd('git') failed due to: exit code(1)                                                                                           run.py:138
                      cmdline: git commit -m upload pdf report                                                                                                                     
                      stdout: 'On branch osa_tool_attachments                                                                                                                      
                    nothing to commit, working tree clean'                                                                                                                         
                    Traceback (most recent call last):                                                                                                                             
                      File "/home/ilya/OSA/osa_tool/run.py", line 98, in main                                                                                                      
                        github_agent.upload_report(analytics.filename, analytics.output_path)                                                                                      
                      File "/home/ilya/OSA/osa_tool/github_agent/github_agent.py", line 280, in upload_report                                                                      
                        self.commit_and_push_changes(report_branch, commit_message, True)                                                                                          
                      File "/home/ilya/OSA/osa_tool/github_agent/github_agent.py", line 185, in commit_and_push_changes                                                            
                        self.repo.git.commit('-m', commit_message)                                                                                                                 
                      File "/home/ilya/.cache/pypoetry/virtualenvs/osa-tool-hFIV0qD6-py3.10/lib/python3.10/site-packages/git/cmd.py", line 986, in <lambda>                        
                        return lambda *args, **kwargs: self._call_process(name, *args, **kwargs)                                                                                   
                      File "/home/ilya/.cache/pypoetry/virtualenvs/osa-tool-hFIV0qD6-py3.10/lib/python3.10/site-packages/git/cmd.py", line 1598, in _call_process                  
                        return self.execute(call, **exec_kwargs)                                                                                                                   
                      File "/home/ilya/.cache/pypoetry/virtualenvs/osa-tool-hFIV0qD6-py3.10/lib/python3.10/site-packages/git/cmd.py", line 1388, in execute                        
                        raise GitCommandError(redacted_command, status, stderr_value, stdout_value)                                                                                
                    git.exc.GitCommandError: Cmd('git') failed due to: exit code(1)                                                                                                
                      cmdline: git commit -m upload pdf report                                                                                                                     
                      stdout: 'On branch osa_tool_attachments                                                                                                                      
                    nothing to commit, working tree clean' 
```
Image

## Cause

pdf files are in .gitignore

## Suggested Solution

Catch the error and handle it with appropriate log message.

## Additional Information

Found by @Ekarerina-Ulyanova

### PR DESCRIPTION
**Pull request by DRMPN**
- [x] Fix report uploading:
  - [x] Repo just downloaded, but fork exists
  - [x] Checkout to report branch when re-run multiple times
  - [x] Refactor if needed
- [x] Update tests

Closes #149 
Closes #165 

This pull request introduces enhancements to the `osa_tool` codebase, focusing on improving the functionality and robustness of the `upload_report` and `commit_and_push_changes` methods. Additionally, it updates unit tests to ensure coverage for the new features and edge cases. 

### Feature Enhancements:

* [`osa_tool/github_agent/github_agent.py`](diffhunk://#diff-7eb2846b6fd626256df7a6b3145bc1c5d3f68ecb1e69b9bc6fb234dff4338105L183-R186): Added a `force` parameter to the `commit_and_push_changes` method, allowing forced pushes when necessary. Updated the `upload_report` method to handle cases where the report branch already exists and to move the report file safely between branches. [[1]](diffhunk://#diff-7eb2846b6fd626256df7a6b3145bc1c5d3f68ecb1e69b9bc6fb234dff4338105L183-R186) [[2]](diffhunk://#diff-7eb2846b6fd626256df7a6b3145bc1c5d3f68ecb1e69b9bc6fb234dff4338105L205-R216) [[3]](diffhunk://#diff-7eb2846b6fd626256df7a6b3145bc1c5d3f68ecb1e69b9bc6fb234dff4338105R287-R310)

* [`osa_tool/run.py`](diffhunk://#diff-bfb7a699dcee695a9c934b483c7070fe067ac5dd62cdb725bb055386592dc0d0L99-R99): Modified the `upload_report` method call to include the report's file path, ensuring proper handling of the report file during uploads.

### Unit Test Updates:

* [`tests/unit/github_agent/test_commit_push.py`](diffhunk://#diff-b0492e73b65425d4726192833634b655d1fbcba1fefc7def2904a11327b786ceR1-L2): Updated the `test_commit_and_push_changes_success` test to validate the behavior of the new `force` parameter in the `commit_and_push_changes` method. [[1]](diffhunk://#diff-b0492e73b65425d4726192833634b655d1fbcba1fefc7def2904a11327b786ceR1-L2) [[2]](diffhunk://#diff-b0492e73b65425d4726192833634b655d1fbcba1fefc7def2904a11327b786ceL23-R24)

* [`tests/unit/github_agent/test_upload_report.py`](diffhunk://#diff-854d7258355a338e4ff318db8c7f411ed04227933ac8b5f6f561cc84f6467f0dR4-R15): Added new tests to cover scenarios such as commit failures due to `.gitignore` restrictions and handling existing branches during report uploads. Updated existing tests to accommodate the `report_filepath` parameter and the `force` option in pushes. [[1]](diffhunk://#diff-854d7258355a338e4ff318db8c7f411ed04227933ac8b5f6f561cc84f6467f0dR4-R15) [[2]](diffhunk://#diff-854d7258355a338e4ff318db8c7f411ed04227933ac8b5f6f561cc84f6467f0dR40-R68) [[3]](diffhunk://#diff-854d7258355a338e4ff318db8c7f411ed04227933ac8b5f6f561cc84f6467f0dL52-R78) [[4]](diffhunk://#diff-854d7258355a338e4ff318db8c7f411ed04227933ac8b5f6f561cc84f6467f0dL81-R112)

### CODE REVIEW

**Comment by DRMPN** [comment]
Нашел и обработал 2 случая:

1.  Случай, когда запускаешь осу в первый раз с скачиванием репозитория, но уже были запуски в прошлом, так что форк и ветка `osa_tool_attachments` уже существуют:
image
2.  Случай, когда репозиторий уже скачан и запускаешь осу многократно:
image
3. При обработке новых репозиториев ничего не изменилось:
image

File: osa_tool/github_agent/github_agent.py [hunk 1]
```diff
@@ -190,14 +190,7 @@ def commit_and_push_changes(self, branch: str = None, commit_message: str = "osa
...
         except GitCommandError as e:
-            logger.error(
-                f"""Push failed: Branch '{branch}' already exists in the fork.
-             To resolve this, please either:
-                1. Choose a different branch name that doesn't exist in the fork 
-                   by modifying the `branch_name` parameter.
-                2. Delete the existing branch from forked repository.
-                3. Delete the fork entirely.""")
-            return False
+            raise
```
Comment (nicl-nno) [inline osa_tool/github_agent/github_agent.py:193]: Так а зачем так? Ну не пушнулось и не пушнулось, код выше может сам решить что с этим делать.

Просто вроде операция довольно часто используемая, и получается её везде надо обертывать в try-except, чтобы воркфлоу не ломался.

Comment (DRMPN) [reply osa_tool/github_agent/github_agent.py]: Поправил, теперь работает следующим образом:
image

**Comment by DRMPN** [comment]
Фикс на #165. 

Обработал ошибку и вывел сообщение в логи.

image

**Review by jrzkaminski — APPROVED** [review]

File: osa_tool/github_agent/github_agent.py [hunk 2]
```diff
@@ -276,9 +284,31 @@ def upload_report(
...
+        except GitCommandError:
+            logger.warning(
+                f"PDF report located in the {self.branch_name} branch, moving it."
+            )
+            with open(report_filepath, "rb") as f:
+                preserved_content = f.read()
+            self.repo.git.checkout("-f", report_branch)
+            with open(report_filepath, "wb") as f:
+                f.write(preserved_content)
+            logger.info(f"Sucessfuly moved PDF report to {report_branch} branch.")
```
[the remark this answers is not in the document]
Comment (jrzkaminski) [reply osa_tool/github_agent/github_agent.py]: Норм замечание

**Comment by DRMPN** [comment]
Я добавил новые параметры к функциям, но забыл указать их в docstring. 
С помощью copilot это вспомнил и даже на секунду подумал, что он их добавил, но нет, эйфория оказалась недолгой.

Поправлю.
